### DISCUSSION TITLE
New multiprocess manager

### DISCUSSION THREAD

**Discussion by abersheeran** [Kludex/uvicorn#2164] [Ideas] [answered]
https://github.com/abersheeran/zibai/blob/main/src/zibai/multiprocess.py

I have been using the multi-process manager in runweb for a while now, and it works really well. If there are plans to optimize uvicorn's current multi-process setup, I would be happy to submit a pull request and contribute my experience to uvicorn.

**Comment by abersheeran** [comment]
cc @encode/maintainers

**Comment by Kludex** [comment, marked as answer]
Please do. If you can explain everything on the PR, it would be great. 🙏

**Comment by ahopkins** [comment]
Chiming in to also hear what is planned and offer advice. I went down this route fairly recently as well.

**Comment by lmachineone** [comment]
is this a replacement for gunicorn + uvicorn? (meaning we could achieve multiple workers better with just uvicorn now?)
How does this compare over using gunicorn + uvicorn?

**Reply by Kludex** [reply]
Yes.

**Reply by adriangb** [reply]
Where are you deploying that you have multiple CPUs available?

**Comment by lmachineone** [comment]
Thanks, will test this to compare performance some moment

**Comment by lmachineone** [comment]
I only have multiple cpu cores, not cpus itself

**Reply by adriangb** [reply]
But is this a bare metal server, an EC2 instance, etc?

**Reply by lmachineone** [reply]
EC2, t4g.medium, basically, 2 shared cores of an ARM CPU

**Reply by adriangb** [reply]
Ok makes sense that you'd want to use multiprocessing then! I just want to make sure you've considered the option of scaling horizontally via your infrastructure provider.

**Reply by lmachineone** [reply]
That's very kind of you! Thank you! Yeah, I am in a very early startup, but hopefully in the future I will need to start considering horizontal scaling too :)

**Comment by philipbjorge** [comment]
> How does this compare over using gunicorn + uvicorn?

@Kludex -- 
Are you able to comment on this question?
We're trying to figure out whether we want to swap to the uvicorn process manager or continue using the gunicorn + uvicorn workers which has been working well for us.
